src/mame/drivers/enmirage.cpp

```diff
@@ -103,109 +165,261 @@ void enmirage_state::floppy_formats(format_registration &fr)
 }
 
+void enmirage_state::coefficients_w(offs_t offset, uint8_t data )
+{
+	uint8_t channel = offset & 0x07;
+	uint8_t filter_input = (offset >> 3) & 0x03;
+
+	LOGFILTERWRITE("%s, filter update: channel: %d, data: $%02x (%s%s%s%s)\n",
+				machine().describe_context(),
+				channel,
+				data,
+				(filter_input & 0x01) == 0 ? "VF" : "", /* cut-off frequency */
+				(filter_input & 0x03) == 0 ? " and " : "",
+				(filter_input & 0x02) == 0 ? "VQ" : "", /* filter resonance */
+				(filter_input & 0x03) == 0x03 ? "preload dac" : "" );
+}
+
+// port A:
+//	bits 5/6/7 keypad rows 0/1/2 return
+INPUT_CHANGED_MEMBER(enmirage_state::input_changed)
+{
+	update_keypad_matrix();
+}
+
+void enmirage_state::update_keypad_matrix()
+{
+	uint8_t value;
+
+	value  = ((m_key[0]->read() >> m_key_col_select) & 0x01) << 5;
+	value |= ((m_key[1]->read() >> m_key_col_select) & 0x01) << 6;
+	value |= ((m_key[2]->read() >> m_key_col_select) & 0x01) << 7;
+
+	m_via->write_pa(value);
+}
+
+// port B:
+//	bit 6: IN disk load
+//	bit 5: IN Q Chip sync
+
+uint8_t enmirage_state::mirage_via_read_portb()
+{
+	uint8_t value = m_via->read_pb();
+
+	floppy_image_device *floppy = m_floppy_connector ? m_floppy_connector->get_device() : nullptr;
+	if (floppy)
+	{
+		if (floppy->dskchg_r())
+			value |= 0x40;
+		else
+			value &= ~0x40;
+	}
+
+	return value;
+}
+
 // port A: front panel
-// bits 0-2: column select from 0-7
-// bits 3/4 = right and left LED enable
-// bits 5/6/7 keypad rows 0/1/2 return
+// bits 0/1/2: dual purpose (0 to 7 lines, though a 74LS145 decoder):
+//		keyboard matrix column select
+//		7 segment display driver
+//	bits 3/4 = right and left 7 segment display enable
+//	bits 5/6/7 = Keyboard matrix row sense from 0 to 2
 void enmirage_state::mirage_via_write_porta(uint8_t data)
 {
 	u8 segdata = data & 7;
 	m_display->matrix(((data >> 3) & 3) ^ 3, (1<<segdata));
+
+	uint8_t new_select = (data & 0x07);
+	if (m_key_col_select != new_select)
+	{
+		m_key_col_select = new_select;
+		update_keypad_matrix();
+	}
 }
 
 // port B:
-//  bit 7: OUT UART clock
-//  bit 4: OUT disk enable (motor on?)
-//  bit 3: OUT sample/play
-//  bit 2: OUT mic line/in
-//  bit 1: OUT upper/lower bank (64k halves)
-//  bit 0: OUT bank 0/bank 1 (32k halves)
+//	bit 7: OUT UART clock
+//	bit 4: OUT disk select, motor on, and 6500/11 reset
+//	bit 3: OUT sample/play
+//	bit 2: OUT mic line/in
+//	bit 1: OUT upper/lower bank (64k halves)
+//	bit 0: OUT bank 0/bank 1 (32k quarters)
 
 void enmirage_state::mirage_via_write_portb(uint8_t data)
 {
 	int bank = 0;
 
 	// handle sound RAM bank switching
-	bank = (data & 2) ? (64*1024) : 0;
-	bank += (data & 1) ? (32*1024) : 0;
-	if (bank != last_sndram_bank)
-	{
-		last_sndram_bank = bank;
-		membank("sndbank")->set_base(memregion("es5503")->base() + bank);
-	}
+	bank = data & 0x03;
+	m_sample_bank->set_entry(bank);
+
+	// handle floppy motor on
+	floppy_image_device *flop = m_floppy_connector->get_device();
```

Comment:
In `mirage_via_read_portb` you check that the floppy device is non-null before using it, but you don’t check here.  Is it guaranteed to be non-null, or should it be checked here?

Reply:
It should be checked here.